src/analysis/processing/qgsalgorithmjoinbylocation.cpp

```diff
@@ -0,0 +1,345 @@
+        mUnjoinedFeatures->addFeature( f2, QgsFeatureSink::FastInsert );
+    }
+  }
+
+  QVariantMap outputs;
+  if ( mJoinedFeatures )
+  {
+    outputs.insert( QStringLiteral( "OUTPUT" ), mJoinedFeaturesId );
+  }
+  if ( mUnjoinedFeatures )
+  {
+    outputs.insert( QStringLiteral( "NON_MATCHING" ), mUnjoinedFeaturesId );
+  }
+  outputs.insert( QStringLiteral( "JOINED_COUNT" ), joinedCount );
+  return outputs;
+}
+
+bool QgsJoinByLocationAlgorithm::featureFilter( const QgsFeature &feature, std::unique_ptr< QgsGeometryEngine > engine ) const
+{
+  const QgsAbstractGeometry *geom = feature.geometry().constGet();
+  bool ok = false;
+  for ( const auto  predicate : mPredicates )
+  {
+    switch ( predicate )
+    {
+      case 0:
+        if ( engine->intersects( geom ) )
+        {
+          ok = true;
+        }
+        break;
+      case 1:
+        if ( engine->within( geom ) )
+        {
+          ok = true;
+        }
+        break;
+      case 2:
+        if ( engine->isEqual( geom ) )
+        {
+          ok = true;
+        }
+        break;
+      case 3:
+        if ( engine->touches( geom ) )
+        {
+          ok = true;
+        }
+        break;
+      case 4:
+        if ( engine->overlaps( geom ) )
+        {
+          ok = true;
+        }
+        break;
+      case 5:
+        if ( engine->contains( geom ) )
+        {
+          ok = true;
+        }
+        break;
+      case 6:
+        if ( engine->crosses( geom ) )
+        {
+          ok = true;
+        }
+        break;
+    }
+    if ( ok )
+      return ok;
+  }
+  return ok;
+}
+
+bool QgsJoinByLocationAlgorithm::processFeatures( QgsFeature &joinFeature, QgsProcessingFeedback *feedback )
+{
+
+  if ( !joinFeature.hasGeometry() )
+    return false;
+  const QgsGeometry featGeom = joinFeature.geometry();
+  std::unique_ptr< QgsGeometryEngine > engine = qgis::make_unique< QgsGeometryEngine >( QgsGeometry::createGeometryEngine( featGeom.constGet() ) );
+  engine->prepareGeometry();
+  QgsFeatureRequest req = QgsFeatureRequest().setFilterRect( featGeom.boundingBox() );
+  QgsFeatureIterator it = mBaseSource->getFeatures( req );
+  QList<QgsFeature> filtered;
+  QgsFeature baseFeature;
+  bool ok = false;
+  QgsAttributes joinAttributes;
+  for ( int ix : qgis::as_const( mFields2Indices ) )
+  {
+    joinAttributes.append( joinFeature.attribute( ix ) );
+  }
+
+  while ( it.nextFeature( baseFeature ) )
+  {
+    if ( feedback->isCanceled() )
+      break;
+    if ( mJoinMethod == 1 && !mUnjoinedIds.contains( baseFeature.id() ) )
+      continue;
+    if ( featureFilter( baseFeature, engine ) )
+    {
+      if ( mJoinedFeatures )
+      {
+        QgsFeature outputFeature( baseFeature );
+        outputFeature.setAttributes( baseFeature.attributes() + joinAttributes );
+        mJoinedFeatures->addFeature( outputFeature, QgsFeatureSink::FastInsert );
+      }
+      if ( !ok )
+        ok = true;
+      if ( !mDiscardNonMatching && mUnjoinedFeatures )
```

Comment:
```suggestion
      if ( !mDiscardNonMatching || mUnjoinedFeatures )
```

Reply:
Are you sure about this? Since there is no more check to make the unjoined output I would assume that there  OR conditions should always be true as the unjoined sink is always created.